File: 15/python/main.py
import math
import heapq


def shortest_path(grid):
    def unvisited_neighbors(row, col):
        def in_range(r, c):
            return r >= 0 and c >= 0 and r < len(grid) and c < len(grid[0])

        return (
            next_coord
            for next_coord in (
                (row + 1, col),
                (row - 1, col),
                (row, col + 1),
                (row, col - 1),
            )
            if next_coord in unvisited and in_range(*next_coord)
        )

    n_rows, n_cols = len(grid), len(grid[0])
    unvisited = {(row, col) for row in range(n_rows) for col in range(n_cols)}
    distances = {(r, c): math.inf for r in range(n_rows) for c in range(n_cols)}
    distances[(0, 0)] = 0
    heap = []
    heapq.heappush(heap, (0, (0, 0)))
    while unvisited:
        dist_to_curr, curr = heapq.heappop(heap)
        if curr not in unvisited:
            continue
        unvisited.remove(curr)
        for r, c in unvisited_neighbors(*curr):
            total_dist = dist_to_curr + grid[r][c]
            if total_dist < distances[(r, c)]:
                distances[(r, c)] = total_dist
                heapq.heappush(heap, (total_dist, (r, c)))

    return distances[(n_rows - 1, n_cols - 1)]


def make_big_grid(grid):
    n_rows, n_cols = len(grid), len(grid[0])
    result = [[0] * 5 * n_cols for _ in range(5 * n_rows)]
    for r in range(len(grid)):
        for c in range(len(grid)):
            for i in range(5):
                for j in range(5):
                    val = (grid[r][c] + i + j - 1) % 9 + 1
                    result[r + i * n_rows][c + j * n_cols] = val
    return result


with open("../input.txt") as f:
    grid = [[int(c) for c in line.strip()] for line in f]
    print(shortest_path(grid))
    print(shortest_path(make_big_grid(grid)))

File: 15/rust/src/main.rs
use std::{cmp::Reverse, collections::BinaryHeap};

fn main() {
    let grid: Vec<Vec<i32>> = include_str!("../../input.txt")
        .lines()
        .map(|line| {
            line.chars()
                .map(|c| c.to_digit(10).expect("invalid input") as i32)
                .collect()
        })
        .collect();
    let num_rows = grid.len();
    let num_cols = grid[0].len();
    println!(
        "part 1: {}",
        shortest_path(|r, c| grid[r][c], num_rows, num_cols)
    );
    let extended_grid_lookup = |extended_grid_row: usize, extended_grid_col: usize| {
        let (row_offset, row) = (extended_grid_row / num_rows, extended_grid_row % num_rows);
        let (col_offset, col) = (extended_grid_col / num_cols, extended_grid_col % num_cols);
        let val = grid[row][col];
        (val - 1 + row_offset as i32 + col_offset as i32) % 9 + 1
    };
    println!(
        "part 2: {}",
        shortest_path(extended_grid_lookup, 5 * num_rows, 5 * num_cols)
    );
}

fn shortest_path<F>(grid_val: F, num_rows: usize, num_cols: usize) -> i32
where
    F: Fn(usize, usize) -> i32,
{
    let in_range = |row, col| row < num_rows && col < num_cols;
    let neighbors = |(r, c): (usize, usize)| {
        let (r, c) = (r as i32, c as i32);
        [(1, 0), (-1, 0), (0, 1), (0, -1)]
            .into_iter()
            .map(move |(dr, dc)| (r + dr, c + dc))
            .filter(|(r, c)| *r >= 0 && *c >= 0 && *r < num_rows as i32 && *c < num_cols as i32)
            .map(|(r, c)| (r as usize, c as usize))
    };
    let mut distances = vec![vec![i32::MAX; num_cols]; num_rows];
    distances[0][0] = 0;
    let mut pq = BinaryHeap::new();
    pq.push(Reverse((0, (0, 0))));
    loop {
        let Reverse((dist, (row, col))) = pq.pop().expect("illegal state");
        for (r, c) in neighbors((row, col)) {
            if in_range(r, c) {
                let new_dist = grid_val(r, c) + dist;
                let old_dist = &mut distances[r][c];
                if new_dist < *old_dist {
                    *old_dist = new_dist;
                    pq.push(Reverse((new_dist, (r, c))));
                }
            }
        }
        if row == num_rows - 1 && col == num_cols - 1 {
            break distances[row][col];
        }
    }
}
